File: experiments/problem1_nonperiodicity/analyze_period_two_signature_simulation_no_go.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iostream>
#include <map>
#include <optional>
#include <set>
#include <stdexcept>
#include <tuple>
#include <vector>

using u64 = std::uint64_t;
constexpr int MAXK = 25;
constexpr std::array<unsigned, 4> CHILD{0b1011, 0b1100, 0b1110, 0b0011};

u64 forward(char generator, u64 state) {
    const u64 stepped = state ^ ((state << 1) | (state << 2));
    if (generator == 't') return stepped;
    if (generator == 'u') return stepped ^ 1;
    if (generator == 'p') return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    throw std::runtime_error("unknown generator");
}

std::vector<u64> next_level(const std::vector<u64>& current) {
    std::vector<u64> output;
    output.reserve(current.size() * 3);
    for (u64 state : current) {
        output.push_back(forward('t', state));
        output.push_back(forward('u', state));
        output.push_back(forward('p', state));
    }
    std::sort(output.begin(), output.end());
    output.erase(std::unique(output.begin(), output.end()), output.end());
    return output;
}

bool contains(const std::vector<u64>& values, u64 target) {
    return std::binary_search(values.begin(), values.end(), target);
}

int bit_length(u64 value) {
    return value ? 64 - __builtin_clzll(value) : 0;
}

std::optional<u64> inverse_t(u64 output) {
    if (output == 0) return u64{0};
    const int width = bit_length(output) - 2;
    if (width <= 0) return std::nullopt;
    u64 state = 0;
    for (int position = 0; position < width; ++position) {
        unsigned lower = 0;
        if (position >= 1) lower |= (state >> (position - 1)) & 1;
        if (position >= 2) lower |= (state >> (position - 2)) & 1;
        const unsigned bit = ((output >> position) & 1) ^ lower;
        state |= u64(bit) << position;
    }
    if (forward('t', state) != output) return std::nullopt;
    return state;
}

std::optional<u64> inverse_generator(char generator, u64 output) {
    std::optional<u64> state;
    if (generator == 't') {
        state = inverse_t(output);
    } else if (generator == 'u') {
        state = inverse_t(output ^ 1);
    } else if (generator == 'p') {
        const unsigned low = (output & 1) ^ 1;
        state = inverse_t(output ^ 1 ^ (low == 0 ? 2 : 0));
    } else {
        throw std::runtime_error("unknown generator");
    }
    if (!state || forward(generator, *state) != output) return std::nullopt;
    return state;
}

std::optional<u64> candidate_parent(u64 quotient, int digit) {
    const char generator = digit == 0 ? 't' : digit == 1 ? 'u' : 'p';
    const auto residual = inverse_generator(generator, quotient);
    if (!residual) return std::nullopt;
    return 4 * (*residual) + unsigned(digit);
}

unsigned fiber_from_predecessor(unsigned predecessor) {
    unsigned fiber = 0;
    for (int digit = 0; digit < 4; ++digit) {
        if ((predecessor >> digit) & 1) fiber |= CHILD[digit];
    }
    return fiber;
}

unsigned predecessor_mask(const std::vector<u64>& level, u64 quotient) {
    unsigned predecessor = 0;
    for (int digit = 0; digit < 4; ++digit) {
        const auto parent = candidate_parent(quotient, digit);
        if (parent && contains(level, *parent)) predecessor |= 1u << digit;
    }
    if ((predecessor & 4) && !(predecessor & 8)) {
        throw std::runtime_error("digit-2 predecessor lacked digit-3 mate");
    }
    return predecessor;
}

unsigned signature(const std::vector<u64>& level, u64 state) {
    const unsigned predecessor = predecessor_mask(level, state);
    return (predecessor << 4) | fiber_from_predecessor(predecessor);
}

struct Profile {
    unsigned source;
    std::array<unsigned, 4> target;
    bool operator<(const Profile& other) const {
        return std::tie(source, target) < std::tie(other.source, other.target);
    }
};

struct Campaign {
    u64 outputs = 0;
    std::set<unsigned> signatures;
    std::set<u64> edges;
    std::set<Profile> profiles;
    std::map<int, int> profiles_by_level;
};

Campaign campaign(char phase) {
    std::vector<u64> current{phase == 'p' ? 3ULL : 1ULL};
    Campaign result;
    for (int complexity = 1; complexity <= MAXK; ++complexity) {
        result.outputs += current.size();
        std::vector<unsigned> current_signatures;
        current_signatures.reserve(current.size());
        for (u64 state : current) {
            const unsigned value = signature(current, state);
            current_signatures.push_back(value);
            result.signatures.insert(value);
        }
        if (complexity == MAXK) break;

        auto next = next_level(current);
        std::vector<unsigned> next_signatures;
        next_signatures.reserve(next.size());
        for (u64 state : next) {
            const unsigned value = signature(next, state);
            next_signatures.push_back(value);
            result.signatures.insert(value);
        }

        std::set<Profile> level_profiles;
        for (std::size_t index = 0; index < current.size(); ++index) {
            const u64 state = current[index];
            Profile profile;
            profile.source = current_signatures[index];
            profile.target.fill(0xFFFF);
            for (int digit = 0; digit < 4; ++digit) {
                const u64 child = 4 * state + digit;
                const auto found = std::lower_bound(next.begin(), next.end(), child);
                if (found == next.end() || *found != child) continue;
                const std::size_t child_index = found - next.begin();
                const unsigned target = next_signatures[child_index];
                profile.target[digit] = target;
                const u64 code = (u64(profile.source) << 24)
                    | (u64(digit) << 20) | u64(target);
                result.edges.insert(code);
            }
            result.profiles.insert(profile);
            level_profiles.insert(profile);
        }
        result.profiles_by_level[complexity] = level_profiles.size();
        current.swap(next);
    }
    return result;
}

int main() {
    const Campaign phase_p = campaign('p');
    const Campaign phase_u = campaign('u');
    std::vector<unsigned> signatures(
        phase_p.signatures.begin(), phase_p.signatures.end()
    );
    std::map<unsigned, int> index;
    for (int position = 0; position < static_cast<int>(signatures.size()); ++position) {
        index[signatures[position]] = position;
    }

    const int signature_count = signatures.size();
    std::vector<std::array<unsigned, 4>> post(signature_count);
    for (auto& row : post) row.fill(0);
    for (u64 code : phase_p.edges) {
        const unsigned source = (code >> 24) & 0xFF;
        const unsigned digit = (code >> 20) & 0xF;
        const unsigned target = code & 0xFFFF;
        post[index[source]][digit] |= 1u << index[target];
    }

    const int subset_count = 1 << signature_count;
    std::vector<std::array<unsigned, 4>> subset_post(subset_count);
    for (auto& row : subset_post) row.fill(0);
    for (int subset = 1; subset < subset_count; ++subset) {
        const int bit = __builtin_ctz(subset);
        const int previous = subset & (subset - 1);
        for (int digit = 0; digit < 4; ++digit) {
            subset_post[subset][digit] =
                subset_post[previous][digit] | post[bit][digit];
        }
    }

    std::vector<char> good(signature_count * subset_count, 1);
    for (int current = 0; current < signature_count; ++current) {
        good[current * subset_count] = 0;
    }
    std::vector<int> removed_rounds;
    while (true) {
        std::vector<int> removed;
        for (int current = 0; current < signature_count; ++current) {
            for (int shadow_set = 1; shadow_set < subset_count; ++shadow_set) {
                const int key = current * subset_count + shadow_set;
                if (!good[key]) continue;
                bool valid = true;
                for (int digit = 0; digit < 4 && valid; ++digit) {
                    const unsigned current_targets = post[current][digit];
                    if (!current_targets) continue;
                    const unsigned shadow_targets = subset_post[shadow_set][digit];
                    if (!shadow_targets) {
                        valid = false;
                        break;
                    }
                    for (unsigned bits = current_targets; bits; bits &= bits - 1) {
                        const int target = __builtin_ctz(bits);
                        if (!good[target * subset_count + shadow_targets]) {
                            valid = false;
                            break;
                        }
                    }
                }
                if (!valid) removed.push_back(key);
            }
        }
        if (removed.empty()) break;
        for (int key : removed) good[key] = 0;
        removed_rounds.push_back(removed.size());
    }

    int good_count = 0;
    int singleton_count = 0;
    for (char value : good) good_count += value;
    for (int current = 0; current < signature_count; ++current) {
        for (int shadow = 0; shadow < signature_count; ++shadow) {
            singleton_count += good[current * subset_count + (1 << shadow)];
        }
    }
    const unsigned top_signature = 0xFF;
    const int top_index = index[top_signature];
    int universal_count = 0;
    for (int current = 0; current < signature_count; ++current) {
        universal_count += good[current * subset_count + (1 << top_index)];
    }

    std::vector<u64> phase_p_level_1{3};
    const auto phase_p_level_2 = next_level(phase_p_level_1);
    const unsigned source_signature = signature(phase_p_level_2, 12);
    bool concrete_shadow = false;
    for (u64 state : phase_p_level_1) {
        if ((state & 3) == (12 & 3)) concrete_shadow = true;
    }
    const bool abstract_shadow =
        good[index[source_signature] * subset_count + (1 << top_index)];

    if (signature_count != 12 || phase_p.edges.size() != 194
        || phase_u.edges.size() != 194 || phase_p.edges != phase_u.edges) {
        throw std::runtime_error("signature graph totals changed");
    }
    if (good_count != 47393
        || removed_rounds != std::vector<int>({147, 464, 1136})
        || singleton_count != 54 || universal_count != 12) {
        throw std::runtime_error("fixed point totals changed");
    }
    if (phase_p.profiles.size() != 448 || phase_u.profiles.size() != 441) {
        throw std::runtime_error("profile totals changed");
    }
    if (source_signature != 0x2C || !abstract_shadow || concrete_shadow) {
        throw std::runtime_error("concrete lift counterexample changed");
    }

    std::cout << "maximum_complexity=" << MAXK << '\n';
    std::cout << "phase_p_outputs=" << phase_p.outputs << '\n';
    std::cout << "phase_u_outputs=" << phase_u.outputs << '\n';
    std::cout << "total_outputs=" << phase_p.outputs + phase_u.outputs << '\n';
    std::cout << "signatures=" << signature_count << '\n';
    std::cout << "phase_p_edges=" << phase_p.edges.size() << '\n';
    std::cout << "phase_u_edges=" << phase_u.edges.size() << '\n';
    std::cout << "phase_p_profiles=" << phase_p.profiles.size() << '\n';
    std::cout << "phase_u_profiles=" << phase_u.profiles.size() << '\n';
    std::cout << "good_pairs=" << good_count << '\n';
    std::cout << "all_pairs=" << signature_count * (subset_count - 1) << '\n';
    std::cout << "removed_rounds=";
    for (std::size_t index = 0; index < removed_rounds.size(); ++index) {
        if (index) std::cout << ',';
        std::cout << removed_rounds[index];
    }
    std::cout << '\n';
    std::cout << "singleton_simulations=" << singleton_count << '\n';
    std::cout << "top_signature_simulates=" << universal_count << '\n';
    std::cout << "top_signature=1111/1111\n";
    std::cout << "concrete_source=phase-p/k2/state-12/signature-0010-1100\n";
    std::cout << "abstract_top_shadow=1\n";
    std::cout << "same_digit_previous_shadow=0\n";
    std::cout << "profiles_by_level=";
    for (const auto& [complexity, count] : phase_p.profiles_by_level) {
        std::cout << complexity << ':' << count << ',';
    }
    std::cout << '\n';
}

File: experiments/problem1_nonperiodicity/analyze_period_two_signature_simulation_no_go.py
#!/usr/bin/env python3
"""Greatest set-valued simulation on the period-two signature graph.

The twelve parent/fiber signatures form a finite nondeterministic labelled graph
when concrete frontier states are quotiented by signature. This analyzer
computes the greatest same-digit simulation between one current signature and a
nonempty set of shadow signatures. It then exhibits why that abstraction is
too coarse for survivor-cylinder shadows: the greatest simulation has a
universal singleton shadow, while a small exact frontier cylinder has no
adjacent concrete shadow at all.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from typing import Any

PHASES = ("p", "u")
CHILD_MASK = {0: 0b1011, 1: 0b1100, 2: 0b1110, 3: 0b0011}
DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 20
TOP_SIGNATURE = (0b1111, 0b1111)


class SimulationCampaignLimitError(RuntimeError):
    """Raised before a controlled campaign exceeds its configured cap."""


def forward_generator(name: str, state: int) -> int:
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    return tuple(sorted({forward_generator(name, state) for name in "tup"}))


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def build_levels(phase: str, maximum_complexity: int) -> list[set[int]]:
    levels: list[set[int]] = [set(), {phase_start(phase)}]
    for _ in range(2, maximum_complexity + 1):
        levels.append(
            {child for state in levels[-1] for child in frontier_children(state)}
        )
    return levels


def fiber_from_predecessor(mask: int) -> int:
    fiber = 0
    for digit in range(4):
        if (mask >> digit) & 1:
            fiber |= CHILD_MASK[digit]
    return fiber


def signature_map(level: set[int]) -> dict[int, tuple[int, int]]:
    """Compute all signatures by the exact direct predecessor contribution."""
    predecessor: dict[int, int] = {}
    for parent in level:
        digit = parent & 3
        residual = parent >> 2
        generator = "t" if digit == 0 else "u" if digit == 1 else "p"
        quotient = forward_generator(generator, residual)
        predecessor[quotient] = predecessor.get(quotient, 0) | (1 << digit)
    result: dict[int, tuple[int, int]] = {}
    for state in level:
        mask = predecessor.get(state, 0)
        if mask & 0b0100 and not mask & 0b1000:
            raise AssertionError("digit-2 predecessor lacked digit-3 mate")
        result[state] = (mask, fiber_from_predecessor(mask))
    return result


def format_signature(signature: tuple[int, int]) -> str:
    return f"0b{signature[0]:04b}/0b{signature[1]:04b}"


def phase_graph(phase: str, maximum_complexity: int) -> dict[str, Any]:
    levels = build_levels(phase, maximum_complexity)
    maps = [dict() for _ in range(maximum_complexity + 1)]
    for complexity in range(1, maximum_complexity + 1):
        maps[complexity] = signature_map(levels[complexity])

    signatures: set[tuple[int, int]] = set()
    edges: set[tuple[tuple[int, int], int, tuple[int, int]]] = set()
    profiles: set[
        tuple[tuple[int, int], tuple[tuple[int, int] | None, ...]]
    ] = set()
    profiles_by_level: dict[str, int] = {}

    for complexity in range(1, maximum_complexity):
        current = maps[complexity]
        next_map = maps[complexity + 1]
        level_profiles = set()
        signatures.update(current.values())
        signatures.update(next_map.values())
        for state, source in current.items():
            targets: list[tuple[int, int] | None] = []
            for digit in range(4):
                target = next_map.get(4 * state + digit)
                targets.append(target)
                if target is not None:
                    edges.add((source, digit, target))
            profile = (source, tuple(targets))
            profiles.add(profile)
            level_profiles.add(profile)
        profiles_by_level[str(complexity)] = len(level_profiles)

    return {
        "phase": phase,
        "outputs_checked": sum(
            len(levels[complexity])
            for complexity in range(1, maximum_complexity + 1)
        ),
        "signatures": signatures,
        "edges": edges,
        "concrete_transition_profiles": len(profiles),
        "profiles_by_level": profiles_by_level,
        "levels": levels,
        "maps": maps,
    }


def greatest_set_simulation(
    signatures: set[tuple[int, int]],
    edges: set[tuple[tuple[int, int], int, tuple[int, int]]],
) -> dict[str, Any]:
    ordered = sorted(signatures)
    index = {signature: position for position, signature in enumerate(ordered)}
    state_count = len(ordered)
    subset_count = 1 << state_count

    post = [[0] * 4 for _ in ordered]
    for source, digit, target in edges:
        post[index[source]][digit] |= 1 << index[target]

    subset_post = [[0] * 4 for _ in range(subset_count)]
    for subset in range(1, subset_count):
        low = subset & -subset
        position = low.bit_length() - 1
        previous = subset ^ low
        for digit in range(4):
            subset_post[subset][digit] = (
                subset_post[previous][digit] | post[position][digit]
            )

    good = {
        (current, shadow_set)
        for current in range(state_count)
        for shadow_set in range(1, subset_count)
    }
    removed_rounds: list[int] = []
    while True:
        removed: list[tuple[int, int]] = []
        for current, shadow_set in good:
            valid = True
            for digit in range(4):
                current_targets = post[current][digit]
                if not current_targets:
                    continue
                shadow_targets = subset_post[shadow_set][digit]
                if not shadow_targets:
                    valid = False
                    break
                remaining = current_targets
                while remaining:
                    low = remaining & -remaining
                    target = low.bit_length() - 1
                    remaining ^= low
                    if (target, shadow_targets) not in good:
                        valid = False
                        break
                if not valid:
                    break
            if not valid:
                removed.append((current, shadow_set))
        if not removed:
            break
        for pair in removed:
            good.remove(pair)
        removed_rounds.append(len(removed))

    singleton_rows: dict[str, list[str]] = {}
    singleton_count = 0
    for current, signature in enumerate(ordered):
        simulators = []
        for shadow, shadow_signature in enumerate(ordered):
            if (current, 1 << shadow) in good:
                simulators.append(format_signature(shadow_signature))
                singleton_count += 1
        singleton_rows[format_signature(signature)] = simulators

    top_index = index[TOP_SIGNATURE]
    top_simulated = [
        format_signature(ordered[current])
        for current in range(state_count)
        if (current, 1 << top_index) in good
    ]
    return {
        "signature_count": state_count,
        "nonempty_shadow_subsets": subset_count - 1,
        "all_candidate_pairs": state_count * (subset_count - 1),
        "greatest_fixed_point_pairs": len(good),
        "removed_by_round": removed_rounds,
        "fixed_point_rounds": len(removed_rounds),
        "singleton_simulations": singleton_count,
        "singleton_simulators": singleton_rows,
        "universal_singleton": format_signature(TOP_SIGNATURE),
        "universal_singleton_simulates": top_simulated,
        "ordered_signatures": [format_signature(row) for row in ordered],
        "good": good,
        "index": index,
    }


def concrete_lift_failure(
    phase_row: dict[str, Any], simulation: dict[str, Any]
) -> dict[str, Any]:
    levels = phase_row["levels"]
    maps = phase_row["maps"]
    current = 12
    current_complexity = 2
    depth = 1
    residue = current % (4**depth)
    source_signature = maps[current_complexity][current]
    shadows = sorted(
        state
        for state in levels[current_complexity - 1]
        if state % (4**depth) == residue
    )
    source_index = simulation["index"][source_signature]
    top_index = simulation["index"][TOP_SIGNATURE]
    abstract_good = (source_index, 1 << top_index) in simulation["good"]
    if source_signature != (0b0010, 0b1100):
        raise AssertionError("small source signature changed")
    if not abstract_good or shadows:
        raise AssertionError("concrete abstraction counterexample failed")
    return {
        "phase": "p",
        "current_complexity": current_complexity,
        "current_state": current,
        "current_state_hex": hex(current),
        "current_signature": format_signature(source_signature),
        "cylinder_depth": depth,
        "cylinder_residue": residue,
        "previous_frontier": sorted(levels[current_complexity - 1]),
        "same_cylinder_previous_states": shadows,
        "abstract_universal_shadow": format_signature(TOP_SIGNATURE),
        "abstract_simulation_accepts": abstract_good,
        "consequence": (
            "Signature simulation does not imply a same-cylinder adjacent frontier "
            "shadow because the quotient forgets residue and concrete realization."
        ),
    }


def strip_internal(row: dict[str, Any]) -> dict[str, Any]:
    return {
        key: value
        for key, value in row.items()
        if key not in {"signatures", "edges", "levels", "maps"}
    }


def run_campaign(
    maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY,
) -> dict[str, Any]:
    if not 8 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise SimulationCampaignLimitError("maximum complexity outside controlled range")

    phases = {
        phase: phase_graph(phase, maximum_complexity) for phase in PHASES
    }
    signatures = phases["p"]["signatures"] | phases["u"]["signatures"]
    edges = phases["p"]["edges"] | phases["u"]["edges"]
    if len(signatures) != 12:
        raise AssertionError("signature alphabet did not close at twelve symbols")

    simulation = greatest_set_simulation(signatures, edges)
    concrete = concrete_lift_failure(phases["p"], simulation)
    simulation_public = {
        key: value
        for key, value in simulation.items()
        if key not in {"good", "index"}
    }
    payload: dict[str, Any] = {
        "status": "finite-greatest-signature-simulation-and-concrete-no-go",
        "maximum_complexity": maximum_complexity,
        "signature_graph": {
            "signatures": len(signatures),
            "phase_p_labelled_edges": len(phases["p"]["edges"]),
            "phase_u_labelled_edges": len(phases["u"]["edges"]),
            "union_labelled_edges": len(edges),
            "phase_graphs_identical": phases["p"]["edges"] == phases["u"]["edges"],
        },
        "theorem": {
            "greatest_set_simulation": (
                "For a finite digit-labelled signature graph, (s,B) survives the "
                "greatest fixed point exactly when every labelled successor of s is "
                "again simulated by the deterministic shadow update Post_d(B)."
            ),
            "abstraction_no_go": (
                "A surviving signature-level simulation does not imply an adjacent "
                "same-cylinder frontier shadow; residue and realization consistency "
                "were discarded by the signature quotient."
            ),
        },
        "simulation": simulation_public,
        "concrete_lift_failure": concrete,
        "phases": {phase: strip_internal(row) for phase, row in phases.items()},
        "scientific_boundary": (
            "The fixed point is exact for the configured finite signature graph, and "
            "the concrete lifting failure is exact. The 194-edge stabilization and "
            "transition-profile growth are finite evidence. This result does not prove "
            "an all-depth concrete simulation, the adjacent-shadow inclusion, phase-"
            "complexity divergence, exclusion of eventual period two, or Rule 30 "
            "center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity", type=int, default=DEFAULT_MAXIMUM_COMPLEXITY
    )
    args = parser.parse_args()
    print(json.dumps(run_campaign(args.maximum_complexity), indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
